data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriters.java

```diff
@@ -0,0 +1,612 @@
+            .setFromLongAndScale(data.unscaledValue().longValueExact(), scale);
+      }
+    }
+  }
+
+  private static class Decimal38Converter implements Converter<BigDecimal> {
+    private static final Converter<BigDecimal> INSTANCE = new Decimal38Converter();
+
+    @Override
+    public Class<BigDecimal> getJavaClass() {
+      return BigDecimal.class;
+    }
+
+    @Override
+    public void addValue(int rowId, BigDecimal data, ColumnVector output) {
+      // TODO: validate precision and scale from schema
+      if (data == null) {
+        output.noNulls = false;
+        output.isNull[rowId] = true;
+      } else {
+        output.isNull[rowId] = false;
+        ((DecimalColumnVector) output).vector[rowId].set(HiveDecimal.create(data, false));
+      }
+    }
+  }
+
+  public static class RecordConverter implements Converter<Record> {
+    private final List<Converter> converters;
+
+    RecordConverter(List<Converter> converters) {
+      this.converters = converters;
+    }
+
+    public List<Converter> converters() {
+      return converters;
+    }
+
+    @Override
+    public Class<Record> getJavaClass() {
+      return Record.class;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void addValue(int rowId, Record data, ColumnVector output) {
+      if (data == null) {
+        output.noNulls = false;
+        output.isNull[rowId] = true;
+      } else {
+        output.isNull[rowId] = false;
+        StructColumnVector cv = (StructColumnVector) output;
+        for (int c = 0; c < converters.size(); ++c) {
+          converters.get(c).addValue(rowId, data.get(c, converters.get(c).getJavaClass()), cv.fields[c]);
+        }
+      }
+    }
+  }
+
+  public static class ListConverter implements Converter<List> {
+    private final Converter children;
+
+    ListConverter(Converter children) {
+      this.children = children;
+    }
+
+    @Override
+    public Class<List> getJavaClass() {
+      return List.class;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void addValue(int rowId, List data, ColumnVector output) {
+      if (data == null) {
+        output.noNulls = false;
+        output.isNull[rowId] = true;
+      } else {
+        output.isNull[rowId] = false;
+        List<Object> value = (List<Object>) data;
+        ListColumnVector cv = (ListColumnVector) output;
+        // record the length and start of the list elements
+        cv.lengths[rowId] = value.size();
+        cv.offsets[rowId] = cv.childCount;
+        cv.childCount += cv.lengths[rowId];
+        // make sure the child is big enough
+        cv.child.ensureSize(cv.childCount, true);
+        // Add each element
+        for (int e = 0; e < cv.lengths[rowId]; ++e) {
+          children.addValue((int) (e + cv.offsets[rowId]), value.get(e), cv.child);
+        }
+      }
+    }
+  }
+
+  public static class MapConverter implements Converter<Map> {
+    private final Converter keyConverter;
```

Comment:
keeping these parameterize on wild card may help u remove the unchecked annotation